sdk/identity/azure-identity/src/client_secret_credential.cpp

```diff
@@ -62,28 +68,39 @@ Azure::Core::Credentials::AccessToken ClientSecretCredential::GetToken(
     Azure::Core::Credentials::TokenRequestContext const& tokenRequestContext,
     Azure::Core::Context const& context) const
 {
-  return m_tokenCredentialImpl->GetToken(context, [&]() {
-    using _detail::TokenCredentialImpl;
-    using Azure::Core::Http::HttpMethod;
+  using _detail::TokenCache;
+  using _detail::TokenCredentialImpl;
 
-    std::ostringstream body;
-    body << m_requestBody;
+  std::string scopesStr;
+  {
+    auto const& scopes = tokenRequestContext.Scopes;
+    if (!scopes.empty())
     {
-      auto const& scopes = tokenRequestContext.Scopes;
-      if (!scopes.empty())
+      scopesStr = TokenCredentialImpl::FormatScopes(scopes, m_isAdfs);
+    }
+  }
+
+  return TokenCache::GetToken(m_tenantId, m_clientId, m_authorityHost, scopesStr, [&]() {
+    return m_tokenCredentialImpl->GetToken(context, [&]() {
```

Comment:
Correct me if I'm wrong.

The lambda function defined at Line 84 is invoked when token needs to be refreshed, by the time of which, variable `scopesStr` is already destructed because it's a local variable.

But you capture this local variable by reference rather than by value, and use this variable at Line 92, won't this crash the application?

Reply:
This comment also applies to the code in `managed_identity_source.cpp`

Reply:
No, it shouldn't be dead. The reference to the lambda is not kept anywhere. By the time we're about to return the value, `TokenCache::GetToken()` has already returned the value. It had a chance to call lambda only while it was executing. So by the time we get a return value, we're back in this function, where `scopesStr` is still alive (BTW, should've there been a problem, there would've been the same problem with `m_tokenCredentialImpl` reference).

Simplified, it is this:
```cpp
int f(std::function<int()> g)
{
    return g();
}

int h()
{
    std::string s = "Hello";
    int retval = f([&](){ std::cout << s; return 1; });
    // at this point we returned from f() and from the lambda we passed there.
   // s is stil alive.
   return retval;
}
```

`int retval = ...; return retval` is basically what the compiler generates anyway. Correct, or am I missing something?

Reply:
Ok, got it. So the lambda function is called synchronously. It's not saved somewhere and get called later. Thanks for the clarification!